proposals/0000-pass-array-as-varargs.md

```diff
@@ -0,0 +1,155 @@
+
+* New contextual conversion diagnostics for passing an array to a variadic argument which offer to add the `as T...`.
+
+* A diagnostic for passing regular variadic arguments alongside coerced array elements.
+
+* Diagnostics to enforce the restriction that `as T...` only appears as a function or subscript argument.
+
+### Alternate Spellings
+
+The main alternative to the proposed `as T...` syntax is to use a pound-prefixed keyword when passing array elements as varargs. Some of the potential spellings include `#splat`, `#variadic`, `#passVarargs`, `#asVarargs`, and `#arraySplat`.
+
+## Source compatibility
+
+This proposal is purely additive and has no impact on source compatibility.
+
+## Effect on ABI stability
+
+This proposal does not change the ABI of any existing language features. Variadic arguments are already passed as `Array`s in SIL, so no change is necessary to support the new `as T...` syntax.
+
+## Effect on API resilience
+
+This proposal does not introduce any new features which could become part of a public API.
+
+## Alternatives considered
+
+A number of alternatives to this proposal's approach have been pitched over the years:
+### Implicitly convert `Array`s when passed to variadic arguments
+
+One possibility which has been brought up in the past is abandoning an explicit 'splat' operator in favor of implicitly converting arrays to variadic arguments where possible. It is unclear how this could be implemented without breaking source compatibility. Consider the following example:
+
+```swift
+func f(x: Any...) {}
+f(x: [1, 2, 3]) // Ambiguous!
+```
+In this case, it's ambiguous whether `f(x:)` will receive a single variadic argument of type `[Int]`, or three variadic arguments of type `Int`. This ambiguity arises anytime both the element type `T` and `Array` itself are both convertible to the variadic argument type. It might be possible to introduce default behavior to resolve this ambiguity. However, it would come at the cost of additional complexity, as implicit conversions can be very difficult to reason about. Swift has previously removed similar implicit conversions which could have confusing behavior. It's also worth noting that the standard library's `print` function accepts an `Any...` argument, so this ambiguity would arise fairly often in practice.
+
+It's also unclear how implicit conversions would affect overload ranking, another likely cause of source compatibility breakage.
+
+### Use a leading/trailing `...` or `*` instead of `as T...`
+
+Many past conversations around passing arrays as variadic arguments have pitched using `*` or `...` as 'splat' operators instead of a heavier-weight expression like `as T...`. These operators read clearly at the call site, but they conflict with existing operators in the language. `as T...` has the advantage of avoiding these conflicts, and it's expected that this feature will not be used pervasively throughout a codebase, which helps justify a more verbose spelling.
```

Comment:
Can you provide examples of when ...someArray is malformed or conflicting with existing syntax?

Reply:
I've made some revisions which include editing this section for clarity. In the future, please consider providing proposal feedback on the forums so that others have a chance to comment. Thank you!